### ISSUE TITLE
MSC2716 leftover todo tasks

### ISSUE THREAD

**Issue by MadLittleMods** [matrix-org/synapse#10737]
Just trying to keep track of the last remaining little tasks to get [MSC2716](https://github.com/matrix-org/matrix-doc/pull/2716) good to go.

## Todo

 - [x] Verify `chunk_id_from_query` corresponds to an actual insertion event, [`synapse/rest/client/room_batch.py#L302)`](https://github.com/matrix-org/synapse/blob/d1f1b46c2cfe612790aebdd54765953951c94e45/synapse/rest/client/room_batch.py#L302)
    - -> https://github.com/matrix-org/synapse/pull/10776 
 - [x] Ensure insertion events have unique `next_chunk_id` before persisting. Ensure `(room_id, next_chunk_id)` is unique. Avoid cross-talk/conflicts between chunks
    - -> https://github.com/matrix-org/synapse/pull/10877
    - Reject/soft-fail event over federation
    - Disallow persisting on local server
    - Discussed at https://github.com/matrix-org/synapse/pull/10245#discussion_r669641879
 - [x] Split meta events in `/batch_send` response into their own fields, https://github.com/matrix-org/matrix-doc/pull/2716#discussion_r685508994
    - -> https://github.com/matrix-org/synapse/pull/10777
 - [x] Rename `chunk` to `batch` to match `/batch_send` endpoint, https://github.com/matrix-org/matrix-doc/pull/2716#discussion_r684574497
    - -> https://github.com/matrix-org/synapse/pull/10838
 - [x] Rename `?prev_event` to `?prev_event_id`, https://github.com/matrix-org/matrix-doc/pull/2716#discussion_r705872887
    - -> https://github.com/matrix-org/synapse/pull/10839
 - [ ] ~~https://github.com/matrix-org/synapse/issues/10764~~
     - No longer needed because we now use an empty set of `prev_events` instead of fake `prev_events` 
 - [x] https://github.com/matrix-org/synapse/issues/11091
 - [x] https://github.com/matrix-org/synapse/issues/11241
 - [x] Make "marker" events state events, https://github.com/matrix-org/matrix-doc/pull/2716#discussion_r782629097. When the homeserver sees the marker state change or joins the room again, it can go up the marker event `unsigned.replaces_state` chain of the to collect all of the `m.room.marker` events along the way and set as backward extremities on the homeserver.
    - https://github.com/matrix-org/synapse/pull/12718
    - https://github.com/matrix-org/complement/pull/371
 - [ ] Make Element throw away the timeline and re-fetch when it sees the "marker" state change
    - https://github.com/matrix-org/matrix-react-sdk/pull/8354
    - https://github.com/matrix-org/matrix-js-sdk/pull/2299 
 - [ ] Make Synapse process batches faster
    - -> https://github.com/matrix-org/synapse/pull/11778
 - [ ] Disallow `/batch_send` usage in a room where the room version doesn't support it or is not the room creator
     - Currently, we just don't persist the insertion, batch, and marker events but we shouldn't allow the `/batch_send` usage in the first place to make it more obvious 
     - -> https://github.com/matrix-org/synapse/issues/13216
 - [ ] Add some new background updates to fill in the `insertion_events`, `insertion_event_edges`, `batch_events`, and `insertion_event_extremities` tables when `hs.config.experimental.msc2716_enabled` is enabled so homeservers which go from having the config disabled to enabled(or unsupported to supported) can fill in the necessary info to have historical events appear.
    - As discussed at https://github.com/matrix-org/synapse/pull/10566#discussion_r698992189
    - Other PR where I added a background update, https://github.com/matrix-org/synapse/pull/10697
 - [ ] Optimize [`synapse/handlers/federation_event.py#L545`](https://github.com/matrix-org/synapse/blob/ef7fe09778ad672d6ed80fb2206cfbc11e6a9a5e/synapse/handlers/federation_event.py#L545)(backfill receiving) to be more efficient. Calculate the `stream_ordering` from newest -> oldest (in the correct order) and persist in the oldest -> newest to get the least missing `prev_event` fetch thrashing
    - Spawned from https://github.com/matrix-org/synapse/pull/11114#discussion_r733154176
    - Why aren't we sorting topologically when receiving backfill events? See https://github.com/matrix-org/synapse/pull/11114#discussion_r739610091
    - [x] Blocked on a race condition not being able to run the [Complement test for this](https://github.com/matrix-org/complement/pull/214), https://github.com/matrix-org/synapse/pull/12394#discussion_r872012615
    - [ ] -> https://github.com/matrix-org/synapse/pull/13864 and https://github.com/matrix-org/synapse/pull/13970 and https://github.com/matrix-org/synapse/pull/13971
 - [ ] Figure out how we want to elegantly support relations in the MSC2716 API's. How do I reference a message in the same batch which hasn't been persisted yet.
    - https://github.com/matrix-org/matrix-spec-proposals/pull/2716#discussion_r870884150
 - [ ] (Element problem) When visiting a link to a specific message (permalink), fix messages not loading before it in the scrollback
 - [ ] Write MSC2716 test to see if the bridge can join the room later, then import messages before the room was created
 - [ ] Enable MSC2716 Complement tests for workers

Outstanding PRs that would still be good to merge (even the closed ones):

 - [ ] https://github.com/matrix-org/complement/pull/214
 - [ ] https://github.com/matrix-org/synapse/pull/13229
 - [ ] https://github.com/matrix-org/complement/pull/221

Maybe:   
 
 - [ ] Add  `(room_id, next_batch_id)` unique constraint to DB, https://github.com/matrix-org/synapse/pull/10877#discussion_r716902568
 - [ ] Make `/messages` and `/context` include the state from any historical messages returned. Currently, if the historical message isn't the first or last (depending on which endpoint), the historical state won't be available for the historical messages in the timeline and the avatar/displayname won't show for them.
    - This is a problem for people who individually import historical messages.
    - And is also a problem for when you import a bunch of history to an existing room at the start. The transition between real messages and historical won't have avatar/displaynames shown. Once you reach the next `/messages` back, they will show.
    - (also discussed in the [#10975 PR description](https://github.com/matrix-org/synapse/pull/10975#issue-1013919139))
    - According to the spec, `/messages` returns ["A list of state events relevant to showing the chunk."](https://matrix.org/docs/spec/client_server/r0.6.1#get-matrix-client-r0-rooms-roomid-messages) so we should be able to make this work and also include any historical state associated with some historical messages in the chunk.
    - According to the spec, `/context` only returns ["The state of the room at the last event returned."](https://matrix.org/docs/spec/client_server/r0.6.1#get-matrix-client-r0-rooms-roomid-context-eventid) which doesn't precisely allow wiggle room for adding some random historical state. But we technically could since it's just a list of state events returned.
    - I don't think we have to worry about `/sync` since historical messages will supposedly be way off in the distance and really only fetched by paginating `/messages`. Just to note that `/sync` returns messages for the first message in the chunk, https://github.com/matrix-org/matrix-doc/issues/687
 - [ ] Fix error when using `invite_room_state` on the `/batch_send` state events, as encountered https://github.com/matrix-org/synapse/pull/11220. This isn't any useful functionality gained but would be nice not to throw a 500 here.
 - [ ] Allow state events to be sent in the `events` list
    - The current problem is that the events are persisted in reverse-chronological order so they have the correct stream_ordering while being backfilled which decrements. 
    - As an example of why this fails, if we `/batch_send` -> `events: [messageFromA, memberEventB, messageFromB]`, it will first try to persist `messageFromB` which will fail because `memberEventB` hasn't been processed yet.
    - ["My hunch is that we want to try and persist the entire batch of events all in one go, rather than doing them one-by-one in reverse order."](https://github.com/matrix-org/synapse/pull/12083#discussion_r835172164)
    - See https://github.com/matrix-org/synapse/pull/12083#discussion_r834956170
    - Related https://github.com/matrix-org/synapse/issues/12010
